packages/language-server/src/plugins/prettier/index.ts

```diff
@@ -0,0 +1,174 @@
+import { dirname } from "path";
+import { create as createPrettierService } from "volar-service-prettier";
+import type {
+  FormattingOptions,
+  LanguageServiceContext,
+  LanguageServicePlugin,
+} from "@volar/language-service";
+import { Connection, ShowMessageNotification } from "@volar/language-server";
+import { MessageType } from "@volar/language-server";
+import { URI } from "vscode-uri";
+import { Project } from "@marko/language-tools";
+import * as markoPrettier from "prettier-plugin-marko";
+import type { Options } from "prettier";
+import * as prettierBuiltIn from "prettier";
+import { importMarkoPrettierPlugin, importPrettier } from "./package";
+
+export function createMarkoPrettierService(
+  connection: Connection,
+): LanguageServicePlugin {
+  let prettier: typeof prettierBuiltIn | undefined;
+  let prettierPlugin: typeof markoPrettier | undefined;
+  let hasShownNotification = false;
+
+  return createPrettierService(
+    (context) => {
+      const { prettierInstance, prettierPluginMarko } =
+        getPrettierInstance(context);
+
+      prettier = prettierInstance;
+      prettierPlugin = prettierPluginMarko;
+
+      // Show a warning notification if Prettier or the Marko plugin isn't installed.
+      if ((!prettier || !prettierPlugin) && !hasShownNotification) {
+        connection.sendNotification(ShowMessageNotification.type, {
+          message:
+            "Couldn't load `prettier` or `prettier-plugin-marko`. Falling back to built-in versions.",
+          type: MessageType.Warning,
+        });
+        hasShownNotification = true;
+      }
+
+      if (!prettier) {
+        // Fallback to the built-in version of prettier.
+        prettier = prettierBuiltIn;
+      }
+
+      if (!prettierPlugin) {
+        // Fallback to the built-in version of prettier-plugin-marko.
+        prettierPlugin = markoPrettier;
+      }
```

Comment:
IIRC, this is a change to the existing tools. We use the version from the workspace if it exists and fallback to the one built into the plugin if there isn't one in the workspace.